Add support for additional broker APIs

yolken-segment opened the pull request:
## Description
This change updates the kafka-go `Client` to expose some additional APIs that are needed to support zk-less `topicctl`:

1. `AlterPartitionReassignments`
2. `ApiVersions`
3. `DescribeGroups`
4. `ElectLeaders`
5. `IncrementalAlterConfigs`
6. `ListGroups`

It also fixes the implementation of `DescribeConfigs` when fetching broker configs- based on my testing, I think these need to be split by broker.

## Testing
- Added some unit tests
- Tested the new functionality in the `topicctl` v1 branch: https://github.com/segmentio/topicctl/pull/32

achille-roussel commented on alterpartitionreassignments.go:
```diff
@@ -0,0 +1,121 @@
+	ErrorCode int
+
+	// ErrorMessage describes the top-level error that occured.
+	ErrorMessage string
```
We've expose errors as `error` types in other exported APIs of the client (for example https://pkg.go.dev/github.com/segmentio/kafka-go#FetchResponse), what do you think about continuing this pattern?

(and similarly in other types that include ErrorCode/ErrorMessage)

yolken-segment replied:
Makes sense, done!

achille-roussel commented on describegroups_test.go:
```diff
@@ -48,3 +53,114 @@ func TestDescribeGroupsResponseV0(t *testing.T) {
+		// leads to an error when decoding the DescribeGroupsResponse.
+		//
+		// See https://issues.apache.org/jira/browse/KAFKA-9150 for details.
+		t.Skip("Skipping because kafka version is 2.3.1")
```
Oh Kafka...

achille-roussel commented on protocol/alterpartitionreassignments/alterpartitionreassignments.go:
```diff
@@ -0,0 +1,61 @@
+
+// Detailed API definition: https://kafka.apache.org/protocol#The_Messages_AlterPartitionReassignments
+type Request struct {
+	TimeoutMs int32          `kafka:"min=v0,max=v0"`
+	Topics    []RequestTopic `kafka:"min=v0,max=v0"`
+
+	// We need at least one tagged field to indicate that this is a "flexible" message
+	// type.
+	_ struct{} `kafka:"min=v0,max=v0,tag"`
```
If it doesn't change the semantics, moving this field to the first position will offer a tighter packing in memory:
```suggestion
	// We need at least one tagged field to indicate that this is a "flexible" message
	// type.
	_ struct{} `kafka:"min=v0,max=v0,tag"`

	TimeoutMs int32          `kafka:"min=v0,max=v0"`
	Topics    []RequestTopic `kafka:"min=v0,max=v0"`

```

yolken-segment replied:
Fixed.

achille-roussel commented on protocol/alterpartitionreassignments/alterpartitionreassignments.go:
```diff
@@ -0,0 +1,61 @@
+}
+
+type Response struct {
+	ThrottleTimeMs int32            `kafka:"min=v0,max=v0"`
+	ErrorCode      int16            `kafka:"min=v0,max=v0"`
+	ErrorMessage   string           `kafka:"min=v0,max=v0,nullable"`
+	Results        []ResponseResult `kafka:"min=v0,max=v0"`
+
+	// We need at least one tagged field to indicate that this is a "flexible" message
+	// type.
+	_ struct{} `kafka:"min=v0,max=v0,tag"`
```
If it doesn't change the semantics, moving this field to the first position will offer a tighter packing in memory:

```suggestion
	// We need at least one tagged field to indicate that this is a "flexible" message
	// type.
	_ struct{} `kafka:"min=v0,max=v0,tag"`

	ThrottleTimeMs int32            `kafka:"min=v0,max=v0"`
	ErrorCode      int16            `kafka:"min=v0,max=v0"`
	ErrorMessage   string           `kafka:"min=v0,max=v0,nullable"`
	Results        []ResponseResult `kafka:"min=v0,max=v0"`
```

yolken-segment replied:
Fixed.

nlsun commented on describegroups.go:
```diff
@@ -1,8 +1,147 @@
+		group := DescribeGroupsResponseGroup{
+			GroupID:    apiGroup.GroupID,
+			GroupState: apiGroup.GroupState,
+			Members:    []DescribeGroupsResponseMember{},
```
initializing the slice here and above is usually not necessary

yolken-segment replied:
Fixed.

nlsun commented on describegroups.go:
```diff
@@ -174,3 +313,119 @@ func (t *describeGroupsResponseV0) readFrom(r *bufio.Reader, sz int) (remain int
+	}
+
+	buf := bytes.NewBuffer(rawMetadata)
+	bufReader := bufio.NewReader(buf)
```
you can use the `bytes.Buffer` returned from `bytes.NewBuffer` as a reader below, unless you wanted the buffering here

achille-roussel replied:
Alternatively, `bytes.Reader` with `bytes.NewReader(rawMetadata)` could be an interesting option.

yolken-segment replied:
I think the `read...` helpers we use below specifically require a `bufio.Reader` as opposed to a generic `io.Reader` since they're peeking ahead. Let me know if I'm missing something here?

achille-roussel replied:
You are correct 👍

yolken-segment replied:
Ok :-).

nlsun commented on describegroups.go:
```diff
@@ -174,3 +313,119 @@ func (t *describeGroupsResponseV0) readFrom(r *bufio.Reader, sz int) (remain int
+	mm.Version = int(version16)
+
+	if remain, err = readStringArray(bufReader, remain, &mm.Topics); err != nil {
+		return mm, err
```
might be more clear return the zero value, the `mm` here has `Version` set
```suggestion
		return DescribeGroupsResponseMemberMetadata{}, err
```

yolken-segment replied:
I think this is ok since the caller of this function just discards the decoded metadata if `err` is not `nil`.

nlsun reviewed (commented):
just minor comments, +1 on @achille-roussel 's comments, LGTM otherwise

yolken-segment commented on alterpartitionreassignments_test.go:
```diff
@@ -0,0 +1,58 @@
+	if err != nil {
+		t.Fatal(err)
+	}
+	if resp.ErrorCode != 0 {
+		t.Error(
+			"Bad error code in response",
+			"expected", 0,
+			"got", resp.ErrorCode,
+		)
+	}
```
Yeah, so I deliberately didn't use testify (as much as I'd like to) because it seems like the vast majority of the tests in the repo don't. If it's ok to use testify, maybe we should migrate all of our existing tests in a separate change?

yolken-segment commented on protocol/describeconfigs/describeconfigs.go:
```diff
@@ -16,13 +24,55 @@ type Request struct {
+	messages := []protocol.Message{}
+	topicsMessage := Request{
+		Resources: []RequestResource{},
+	}
```
Fixed.

yolken-segment commented on protocol/incrementalalterconfigs/incrementalalterconfigs.go:
```diff
@@ -0,0 +1,65 @@
+func (r *Request) ApiKey() protocol.ApiKey { return protocol.IncrementalAlterConfigs }
+
+func (r *Request) Broker(cluster protocol.Cluster) (protocol.Broker, error) {
+	// TODO: Support updating multiple broker configs in single request?
```
Done and added a test for this.

yolken-segment commented:
Thanks for the feedback @achille-roussel and @nlsun! I think I addressed all of the feedback, and I also tested the latest changes with `topicctl` v1 to ensure that the new APIs still work.

achille-roussel reviewed (approved):
Changes look good to me 👍 

@nlsun are there more changes you would like to see?

nlsun reviewed (approved):
lgtm
